controller/src/main/java/io/pravega/controller/server/eventProcessor/SerializedRequestHandler.java

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.server.eventProcessor;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.pravega.common.concurrent.FutureHelpers;
+import io.pravega.shared.controller.event.ControllerEvent;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.stream.Collectors;
+
+/**
+ * SerializedRequestHandler class is used to serialize requests for a key and process them.
+ * It maintains a map of key and its work queue.
+ * Any new request received for the key is queued up if queue is not empty. If a worker queue for the key is not found in map,
+ * a new queue is created and the event is put in the queue and this is added to the worker map.
+ * The processing is then scheduled asynchronously for the key.
+ *
+ * Once all pending processing for a key ends, the key is removed from the work map the moment its queue becomes empty.
+ */
+@AllArgsConstructor
+public abstract class SerializedRequestHandler<T extends ControllerEvent> implements RequestHandler<T> {
+
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    private final Map<String, ConcurrentLinkedQueue<Work>> workers = new HashMap<>();
+    private final ExecutorService executor;
+
+    @Override
+    public CompletableFuture<Void> process(final T streamEvent) {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        Work work = new Work(streamEvent, result);
+        String key = streamEvent.getKey();
+
+        synchronized (lock) {
+            if (workers.containsKey(key)) {
+                workers.get(key).add(work);
+            } else {
+                ConcurrentLinkedQueue<Work> queue = new ConcurrentLinkedQueue<>();
+                queue.add(work);
+                workers.put(key, queue);
+                // Start work processing asynchronously and release the lock.
+                executor.execute(() -> run(key, queue));
```

Comment:
You should move this line outside of the lock. There is no need for it to be in the lock. Based on the implementation of the `Executor`, this may block until the queue is available, thus you'll end up holding the lock for a lot longer than you might imagine.

Reply:
done